handler/rfc8693/handler.go

```diff
@@ -0,0 +1,269 @@
+	if !c.CanHandleTokenEndpointRequest(requester) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	if !requester.GetClient().GetGrantTypes().Has(string(fosite.GrantTypeTokenExchange)) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHintf("The OAuth 2.0 Client is not allowed to use authorization grant '%s'.", fosite.GrantTypeTokenExchange))
+	}
+
+	atLifespan := fosite.GetEffectiveLifespan(requester.GetClient(), fosite.GrantTypeTokenExchange, fosite.AccessToken, c.HandleHelper.Config.GetAccessTokenLifespan(ctx))
+
+	if err := c.IssueAccessToken(ctx, atLifespan, requester, responder); err != nil {
+		return err
+	}
+
+	if canIssueRefreshToken(ctx, c, requester) {
+		refresh, refreshSignature, err := c.RefreshTokenStrategy.GenerateRefreshToken(ctx, requester)
+		if err != nil {
+			return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+		}
+		if err := c.RefreshTokenStorage.CreateRefreshTokenSession(ctx, refreshSignature, requester); err != nil {
+			return errorsx.WithStack(fosite.ErrServerError.WithDebug(err.Error()))
+		}
+
+		responder.SetExtra("refresh_token", refresh)
+	}
+	return nil
+}
+
+func canIssueRefreshToken(ctx context.Context, c *Handler, requester fosite.Requester) bool {
+	scope := c.Config.GetRefreshTokenScopes(ctx)
+	// Require one of the refresh token scopes, if set.
+	if len(scope) > 0 && !requester.GetGrantedScopes().HasOneOf(scope...) {
+		return false
+	}
+	// Do not issue a refresh token to clients that cannot use the refresh token grant type.
+	if !requester.GetClient().GetGrantTypes().Has("refresh_token") {
+		return false
+	}
+	return true
+}
+
+func (c *Handler) CanSkipClientAuth(ctx context.Context, requester fosite.AccessRequester) bool {
+	if s := c.Config.GetGrantTypeTokenExchangeCanSkipClientAuth(ctx); s != nil {
+		return s(ctx, requester)
+	}
+
+	return false
+}
+
+func (c *Handler) keyFunc(ctx context.Context) jwt.Keyfunc {
+	return jwt.Keyfunc(func(t *jwt.Token) (interface{}, error) {
+		kid, ok := t.Header["kid"].(string)
+		if !ok {
+			return nil, errors.New("invalid kid")
+		}
+		iss, ok := t.Claims["iss"].(string)
+		if !ok {
+			return nil, errors.New("invalid iss")
+		}
+		return c.Storage.GetIDTokenPublicKey(ctx, iss, kid)
+	})
+}
+
+func (c *Handler) verifyAccessTokenAsSubjectToken(ctx context.Context, clientID string, params *tokenExchangeParams) (fosite.Requester, error) {
+	sig := c.HandleHelper.AccessTokenStrategy.AccessTokenSignature(ctx, params.subjectToken)
+	or, err := c.HandleHelper.AccessTokenStorage.GetAccessTokenSession(ctx, sig, nil)
+	if err != nil {
+		return nil, errorsx.WithStack(fosite.ErrRequestUnauthorized.WithWrap(err).WithDebug(err.Error()))
+	} else if err := c.AccessTokenStrategy.ValidateAccessToken(ctx, or, params.subjectToken); err != nil {
+		return nil, err
+	}
+
+	allowClientIDs, err := c.Storage.GetAllowedClientIDs(ctx, clientID)
```

Comment:
Would it make sense to create a rfc8693.Client interface with `GetAllowedClientIDsForTokenExchange()` instead of putting this in storage? This is going to be part of the client configuration?

Reply:
@vivshankar Sorry I realized there is an error here. It should be 

```go
allowClientIDs, err := c.Storage.GetAllowedClientIDs(ctx, or.GetClient().GetID());
```
The client ID used here should come from the subject token. Because here I want to query the storage for the clients that are allowed to perform token exchange with the given subject token.

In the storage, for each client that is allowing other clients to perform token exchange, we will store a "1 to many" mapping.

Reply:
@saxenautkarsh Thanks. I made a minor change here to use a client function instead -

```
// TokenExchangeAllowed checks if the subject token client allows the specified client
// to perform the exchange
TokenExchangeAllowed(client fosite.Client) bool
```

The reasoning is this gives greater flexibility to the implementer. In reality, if you have a fair number of clients, configuring the matrix of clients is going to be rather difficult. Instead, if a categorizaton mechanism exists (like free-form tags or a `category` property on the client object), that would allow for an easier mechanism to configure this "grouping". It can be implemented in a custom handler certainly but I felt it served the same purpose with more flexibility.